umap/layouts.py

```diff
@@ -346,6 +347,10 @@ def optimize_layout_euclidean(
     if "disable" not in tqdm_kwds:
         tqdm_kwds["disable"] = not verbose
 
+    rng_state_per_sample = np.full(
+        (epochs_per_sample.shape[0], len(rng_state)), rng_state, dtype=np.int64
+    )
+
```

Comment:
You may want adjust this somehow (e.g. add ``np.arange(epochs_per_sample.shape[0])``) so that every point has a slightly different rng_state -- otherwise the negative sample selection is going to be repeated for every sample which may not be ideal.

Reply:
This change would introduce a dependency on row order to transform output, so I would not have the freedom to split my dataset as I would like for the transform to output consistent results regardless of row order. Whilst there are workarounds, I would feel safer if it were completely consistent.

Reply:
I understand; On the other hand the random sampler is now very biased in its sampling, which is a concern. Is there any per point adjustment that might work? The integer part of the initial x-coordinate of the point, for example?

Reply:
That's a good idea. Any number derived from the array will be consistent, but we want it to be generally fast. Perhaps take the x-coordinate (first dimension) and view it as an integer, add to rng_state_per_sample?

Reply:
I would go with the first one; that should be easy enough, and should be consistent.

Reply:
I applied the first option using the head_embedding. Additionally I realised that I should use the head_embedding shape for rng_state_per_sample.